jctools-benchmarks/src/main/java/org/jctools/jmh/latency/QueueBurstCost.java

```diff
@@ -68,102 +71,145 @@ public void setupQueue()
+            {
+                Blackhole.consumeCPU(DELAY_PRODUCER);
+            }
         }
     }
 
-    private void sendBurst(Queue<AbstractEvent> q, int burst, Go go, Stop stop)
+    //do not extends AtomicLong here: JMH will take care of padding fields for us
+    @State(Scope.Thread)
+    public static class Event
     {
-        for (int i = 0; i < burst - 1; i++)
+        private static final long ARRAY_BASE;
+        private static final int ELEMENT_SHIFT;
+
+        static
         {
-            while (!q.offer(go))
+            final int scale = UnsafeAccess.UNSAFE.arrayIndexScale(long[].class);
+            if (8 == scale)
             {
-                ;
+                ELEMENT_SHIFT = 3;
             }
-            if (DELAY_PRODUCER != 0)
+            else
             {
-                Blackhole.consumeCPU(DELAY_PRODUCER);
+                throw new IllegalStateException("Unexpected long[] element size");
             }
+            // Including the buffer pad in the array base offset
+            ARRAY_BASE = UnsafeAccess.UNSAFE.arrayBaseOffset(long[].class);
         }
 
-        while (!q.offer(stop))
+        private static long calcSequenceOffset(long index)
         {
-            ;
+            return ARRAY_BASE + (index << ELEMENT_SHIFT);
         }
-    }
 
-    abstract static class AbstractEvent extends AtomicBoolean
-    {
-        abstract void handle();
-    }
+        private long[] handledCount = null;
 
-    static class Go extends AbstractEvent
-    {
-        @Override
-        void handle()
+        @Setup
+        public void init(QueueBurstCost state)
         {
-            // do nothing
+            handledCount = new long[state.consumerCount];
         }
-    }
 
-    @State(Scope.Thread)
-    public static class Stop extends AbstractEvent
-    {
-        @Override
-        void handle()
+        void reset()
         {
-            lazySet(true);
+            Arrays.fill(handledCount, 0);
+            UnsafeAccess.UNSAFE.storeFence();
+        }
+
+        void waitFor(int value)
+        {
+            final long[] values = this.handledCount;
+            final long length = values.length;
+            do
+            {
+                long total = 0;
+                for (int i = 0; i < length; i++)
+                {
+                    final long offset = calcSequenceOffset(i);
+                    total += UnsafeAccess.UNSAFE.getLongVolatile(values, offset);
+                    if (total == value)
+                    {
+                        return;
+                    }
+                }
+            }
+            while (true);
+        }
+
+        void handle(int consumerId)
+        {
+            final long[] values = this.handledCount;
+            final long offset = calcSequenceOffset(consumerId);
+            final long value = UnsafeAccess.UNSAFE.getLong(values, offset);
+            UnsafeAccess.UNSAFE.putOrderedLong(values, offset, value + 1);
```

Comment:
This introduces significant false sharing on the consumer counters

Reply:
That's correct (as noted by https://github.com/JCTools/JCTools/pull/273#issuecomment-554743228), let me quote myself:
> this could be saved by using 2*CACHE_LINE pads between each counter to avoid consumers invalidating each-others cache-lines while handling producers tasks.

it can works?

Reply:
I've handled false-sharing in the last commit